Weaviate Improvements

trengrj opened the pull request:
We had a look at this benchmark framework and have some improvements in how you are testing Weaviate.

### Things we changed

- Updated Weaviate python client to latest v4 version. This has [numerous performance improvements](https://weaviate.io/blog/grpc-performance-improvements) including adding gRPC support.
- Updated Weaviate to newer [1.24.1 image](https://weaviate.io/blog/weaviate-1-24-release).

Along the way we also fixed isssues in the benchmark including

- Set Weaviate docker config to use the same host’s networking stack as with Qdrant
- Removed unnecessary GC settings
- Improved Weaviate configuration settings in general

### Benchmark reproduction

We reran the benchmark with the same dataset `dbpedia-openai-1M-1536-angular` on same Azure hardware as published:
 
> - Client: 8 vcpus, 16 GiB memory, 64GiB storage (`Standard D8ls v5` on Azure Cloud)
> - Server: 8 vcpus, 32 GiB memory, 64GiB storage (`Standard D8s v3` on Azure Cloud)

Because both vector databases primarily rely on the HNSW algorithm for the vector index, we had a simple goal with this repeat benchmark: Let’s test HNSW performance with as similar parameters as possible.

So instead of creating a comparison with different build parameters (such as on the benchmark site where Qdrant has `efConstruction=512` while Weaviate has `efConstruction=256`) we selected the same build parameters:

- `efConstruction=512` (how deep the graph is searched when indexing a vector)
- `maxConnections=32` (how many edges the graph has on upper levels)

We didn’t use quantization for the following reasons:

- Quantization is impacted heavily by disk caching for performance due to on disk rescoring. This creates large variability especially with the original experiments like `qdrant-sq-rps-m-64-ef-256` that rerun the same queries at different parallel settings.
- We wanted to test base HNSW performance, which is the default algorithm in both databases.

### Results

### Import performance

Database | Configuration | Dataset | Total index time
-- | -- | -- | --
Weaviate | weaviate-1.24.1-m32-efc256 | dbpedia-openai-1M-1536-angular | 20m46s
Qdrant | qdrant-1.8.1-m32-efc256 | dbpedia-openai-1M-1536-angular | 32m02s


Weaviate has <em><strong>37% less total index time</strong></em> for the same HNSW configuration.

Note the benchmark waits until the Qdrant collection is green and HNSW indices merged. Weaviate has an optional async indexing feature that provides similar fast upload times.

### Query performance

![image](https://github.com/qdrant/vector-db-benchmark/assets/2173919/1ee71392-b265-42da-bb9d-8ce1ddfac0d3)

The above graph shows the standard rps/precision curve built by varying the search time `ef` parameter. 

Weaviate is significantly faster 🚀 around the 99% precision target with Weaviate RPS 1225 at 98.9% precision vs Qdrant RPS 471 at 99.0%.

We found the python testing framework on a single instance could not always stress the engine for maximum queries per second (causing the flattened query performance at low `ef`) but agree with the proposal that the majority of users will use a python client.

## Experiment configuration

```json

[{
 "name": "weaviate-1.24.1-m32-efc256",
 "engine": "weaviate",
 "connection_params": {
   "timeout_config": 60
 },
 "collection_params": {
	 "vectorIndexConfig": {
		 "efConstruction": 256,
		 "maxConnections": 32
	 }
 },
 "search_params": [
   { "parallel": 100, "vectorIndexConfig": { "ef": 16} },
   { "parallel": 100, "vectorIndexConfig": { "ef": 32} },
   { "parallel": 100, "vectorIndexConfig": { "ef": 64} },
   { "parallel": 100, "vectorIndexConfig": { "ef": 128} },
   { "parallel": 100, "vectorIndexConfig": { "ef": 256} },
   { "parallel": 100, "vectorIndexConfig": { "ef": 512} },
   { "parallel": 100, "vectorIndexConfig": { "ef": 768} }
 ],
 "upload_params": { "parallel": 8, "batch_size": 1024 }
},
{
  "name": "qdrant-1.8.1-m32-efc256",
  "engine": "qdrant",
  "connection_params": { "timeout": 60 },
  "collection_params": {
    "optimizers_config": { "memmap_threshold":10000000 },
    "hnsw_config": { "m": 32, "ef_construct": 256 }
  },
  "search_params": [
    { "parallel": 100, "search_params": { "hnsw_ef": 16 } },
    { "parallel": 100, "search_params": { "hnsw_ef": 32 } },
    { "parallel": 100, "search_params": { "hnsw_ef": 64 } },
    { "parallel": 100, "search_params": { "hnsw_ef": 128 } },
    { "parallel": 100, "search_params": { "hnsw_ef": 256 } },
    { "parallel": 100, "search_params": { "hnsw_ef": 512 } },
    { "parallel": 100, "search_params": { "hnsw_ef": 768 } }
  ],
  "upload_params": { "parallel": 8, "batch_size": 1024 }
}]
```

As you mentioned that contributions to these are welcome, we would appreciate you merging and updating your benchmark site!

generall commented:
Hi @trengrj, thanks for the contribution!
Happy to see that you finally abandoned ridiculous graphql interface and using something more suitable for database application!

> So instead of creating a comparison with different build parameters (such as on the benchmark site where Qdrant has efConstruction=512 while Weaviate has efConstruction=256) we selected the same build parameters:

I don't think `efConstruction` of qdrant is directly translatable into `efConstruction` of Weaviate, as we don't use a single HNSW graph. For this reason we prefer to compare best-result-config with a given precision.


> We didn’t use quantization for the following reasons:

> Quantization is impacted heavily by disk caching for performance due to on disk rescoring. 

That's up to you, Qdrant can do quantization full in memory

> We wanted to test base HNSW performance, which is the default algorithm in both databases.

At them moment yes, but even with the given algorithm, structure of the storage is pretty different. (Also considering possibility to compare with other engines, which might not necessary use hnsw). We would prefer to keep all possible configurations available as soon as it fits the resource limitations

> As you mentioned that contributions to these are welcome, we would appreciate you merging and updating your benchmark site!

Will definitely do that, but re-running all benchmarks might require some time (we do all runs on single machine to any extra noise from datacenter placement)

generall commented:
Anyway, will keep you updated here on the progress

KShivendu reviewed (commented):
> Weaviate has an optional async indexing feature that provides similar fast upload times.

Import time = Upload time + Indexing time. One needs to wait for indexing to happen. We calculate it the same way for all the engines. 

Looks like the PR was already been merged. I'll create another one to fix the issues that I've pointed out.

etiennedi commented on engine/clients/weaviate/search.py:
```diff
@@ -10,49 +14,42 @@
+        )
+        client.connect()
+        cls.collection = client.collections.get(
+            WEAVIATE_CLASS_NAME, skip_argument_validation=True
```
It's an official recommendation from the [Weaviate docs](https://weaviate.io/developers/weaviate/client-libraries/python#input-argument-validation), which lists considerations of when this should be done. It's a conscious safety&UX vs performance trade-off. The overhead is fairly minimal, but it's not needed when going for performance.

generall commented:
Ok, here is some more detailed breakdown of which configuration are to be used and how to read results.


The main thing is, than qdrant is NOT tuned for high RPS by default. And configuration you used was NOT tuned for high RPS.
Instead it prioritizes low latency and a bit faster indexing.
If you would take a look into latency instead of RPS, you would see completely opposite picture:

![latencies](https://github.com/qdrant/vector-db-benchmark/assets/1935623/ce046896-57b8-4ba6-894c-70ba54dcbe12)

Here is a more detailed explanation why we consider both RPS and Latency-oriented scenarios important:
https://qdrant.tech/benchmarks/#latency-vs-rps

Going back to the RPS-oriented case, have to admit that you made a progress and we are now getting closer to the pure HNSW bottleneck.
Since we started to sqeeze the last bits of performance, here how RPS looks like with a fix more tuned Qdrant configuration:

![vanila-rps](https://github.com/qdrant/vector-db-benchmark/assets/1935623/066d4ebc-62c8-46ba-bbfe-06521c43a699)

This is RPS-oriented config + we allow use of all CPU cores (previously we let one core idle allow to reserve minimal space for other workloads).
Have to admit, I do see Weaviate improves, but not so overwhelming as you would want to think.


On top of that, I would prefer to insist, that limiting comparision to HNSW-only setup limits the potential of the whole technology.
We can go deeper into code-golfing and polish performance of HNSW, that would make vector-db-benchamaring just a less interesting version of the ann-benchmarks. Or, we can be creative and include stuff beyond just happy-path HNSW.

- Big scale
- Distributed deployment
- Filterred/hybrid queries
- Quantizations

For example, that is how result turns if we enable binary quantization:

![bq-rps](https://github.com/qdrant/vector-db-benchmark/assets/1935623/727614ba-5410-468a-a7ec-373852d41ab9)


Looking forward to see your contributions with benchmarking of those new features in Weaviate!
You are stopping us from getting lazy ;)

CC @etiennedi 

---

Raw measurements:

[qdrant-vs-weaviate.zip](https://github.com/qdrant/vector-db-benchmark/files/14625675/qdrant-vs-weaviate.zip)

trengrj commented:
Thanks, looking at the changes:

- You have released a new Qdrant version `1.8.2`  and added custom [environment variables](https://github.com/qdrant/vector-db-benchmark/commit/b4a22b36e187e8ceb9e1be1f6bee842c4ffbe7bd) on the docker image tuned to the test hardware.
- You have switched to two separate configurations, a latency tuned configuration and a rps tuned configuration with different segment values (will just focus on HNSW for now).
- One important thing to note is that Qdrant’s ***total import time is still worse*** compared to Weaviate for both configurations.

I ran some tests with Qdrant `1.8.2` (including the custom environment variables) and continued to use the same Weaviate `1.24.1` configuration.

### Latency

- For the latency graph above you kept the original HNSW configurations unchanged but set `parallel=1` and changed the evaluation metric from requests per seconds (rps) to 95th percentile latency.
- I reran this but also varied `parallel=16` to demonstrate what happens with more concurrency.
- Before looking at latency let’s look at how requests per second performs with these different parallel settings:

![image](https://github.com/qdrant/vector-db-benchmark/assets/2173919/4eee2a52-d948-43da-a887-906fbbe4eeb1)

- With single threaded (`parallel=1`) client performance the result is mixed depending on recall target. However you can see Weaviate scales much better as concurrency is increased from 1 to 16 clients.
- Switching to the 95th percentile latency metric (lower and to the right is better) you can see similar results.

![image](https://github.com/qdrant/vector-db-benchmark/assets/2173919/f40c013a-de24-4700-9c11-e1017b3a6318)

### RPS Configuration

- I next compared the Qdrant rps configuration, again with the same Weaviate configuration.
- The Qdrant rps configuration took much longer to index (over 80% longer) as HNSW segments have to be merged for faster query performance. Weaviate has a mutable HNSW vector index so doesn’t have this problem.

Database | Configuration | Dataset | Total index time
-- | -- | -- | --
Weaviate | weaviate-1.24.1-m32-efc256 | dbpedia-openai-1M-1536-angular | 1254s
Qdrant | qdrant-1.8.2-rps-m32-efc256 | dbpedia-openai-1M-1536-angular | 2268s

- I left Weaviate HNSW build parameters the same, but these could be increased for higher recall (at the cost of slowing down indexing).
- For querying performance at `parallel=16`, Weaviate shows better rps performance at low precision levels and the similar performance at higher precision levels.

![image](https://github.com/qdrant/vector-db-benchmark/assets/2173919/4480634a-db48-4555-9cfa-15b69131cac4)

Included experiment configurations and results are attached.

I do think it is great how we can share these performance findings as open source software. For people who have read this far, I hope this helps for a more realistic assessment of the performance you can get from Weaviate.

[experimental-results.tar.gz](https://github.com/qdrant/vector-db-benchmark/files/14668067/experimental-results.tar.gz)

generall commented:
> The Qdrant rps configuration took much longer to index (over 80% longer) as HNSW segments have to be merged for faster query performance. Weaviate has a mutable HNSW vector index so doesn’t have this problem.

This I can't reproduce on my setup.

For me indexing looks like this:

| Name          | Value |
| ------------- | ----- |
| weaviate      | 1102  |
| qdrant latency| 1121  |
| qdrant rps    | 1366  |
| qdrant rps bq | 730  |

There is slight advantage, but not 80% you claim, and not all all configs.

> I reran this but also varied parallel=16 to demonstrate what happens with more concurrency.

Why would you do that? The point of latency config is to utilize as much resources as possible to execute as fast as possible. Having parallel=16 makes this scenario pointless and basically fallback to RPS contest.

trengrj commented:
> This I can't reproduce on my setup.

Can you supply a more detailed infra setup?

For query performance your most recent graphs in this thread flatline at 1100 requests per second while for both my tests and the original published benchmarks engines can get ~ 1250 for Qdrant (~1370 when I run Weaviate with v4 client).
